lib/puppet/type/mount.rb

```diff
@@ -293,5 +297,18 @@ def value(name)
       dependencies[0..-2]
     end
 
+    # Autorequire the mount point's file resource
+    autorequire(:file) { Pathname.new(@parameters[:name].value) }
+
+    # Autobefore the mount point's child file paths
+    autobefore(:file) do
+      dependencies = []
+      file_resources = catalog.resources.reject { |resource| resource.type != Puppet::Type.type(:file) }
```

Comment:
Also there is a convenient method for looking up a specific resource: `catalog.resource(:file, '/path/to/file')` but it doesn't help the case where you want all resources of a given type.